bgpv2: Allow setting LocalPort from ClusterConfig

YutaroHayakawa opened the pull request:
Currently, `LocalPort` of the `CiliumBGPNodeConfig` is only configurable via `CiliumNodeConfigOverride`. This PR allows configuring it via `CiliumBGPClusterConfig`.

By default, Cilium doesn't listen on the BGP port, because we want to support the setup that Cilium makes a peer with local BGP daemon (listening on 179 may conflict with the local BGP daemon).

However, there's a case, where the peer router doesn't listen on the port and Cilium needs to accept the connection. Instead of creating Override for each node one by one, it's useful to configure the port globally in such a case.

We have discussed out-of-band about naming of the `LocalPort` is potentially confusing (Is it the "source" port number of the locally-initiated connection, or the "listening" port?) and should be `ListenPort`. However, since we are already using this `LocalPort` naming in the NodeConfig and Override. Changing it here makes inconsistency and further confusion. That's why I kept it as is.

```release-note
bgpv2: Allow setting LocalPort from ClusterConfig
```

YutaroHayakawa committed:
bgpv2: Render LocalPort to NodeConfig

When ClusterConfig.Spec.BGPInstances[].LocalPort is specified, render
NodeConfig.Spec.BGPInstances[].LocalPort. The existing Override still
works. When both are specified, override takes precedence.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bgpv2: Add LocalPort in the CiliumBGPClusterConfig

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

harsimran-pabla commented on pkg/k8s/apis/cilium.io/v2alpha1/bgp_cluster_types.go:
```diff
@@ -82,6 +82,15 @@ type CiliumBGPInstance struct {
+	// If not specified, BGP instance will not listen for incoming connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=1
```
Should we add in the description that for port < 1024, cilium would need CAP_NET_BIND_SERVICE privilege?

YutaroHayakawa replied:
Good point. Let me add that.

YutaroHayakawa replied:
Added to the document instead.

YutaroHayakawa committed:
bgpv2,doc: Add localPort description

With caveat that binding on the local BGP port (179) requires additional
capability.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

rastislavs reviewed (approved):
Thanks 👍

YutaroHayakawa commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -291,16 +291,19 @@ func toNodeBGPInstance(clusterBGPInstances []v2alpha1.CiliumBGPInstance, overrid
+				if overrideBGPInstance.LocalPort != nil {
+					nodeBGPInstance.LocalPort = overrideBGPInstance.LocalPort
+				}
 				if overrideBGPInstance.LocalASN != nil {
```
Fixed this case. This should cover the case where the ClusterConfig.LocalPort != nil and Override.LocalPort == nil. In this case, we shouldn't override because it will just clears the global config.
